##### Annotation format announcement

###### For common annotation gff from protein fasta files (from cluster or direct 50 fasta files).

Since multiple lines by different programs are generated by multiple folders, we are making two types of gff files. 

- **GFF Type I** :Annotate results with a **single line**. Different results from different database are annotated in the attribute column of the  final GFF file.

- **GFF Type II**: This output contains multiple records for each gene by multiple sources.



  ```shell
  ID=<Define this after merging>;Target=<Node name>;Each_program_name=Each_match;..;..; 
  ```

###### For merged file

- The common column wiil be defined as follows

  ```shell
  SeqName\tProgram\tFeature\tStart\End\Score\tStrand\tframe\tAttribute	
  ```

- The attribute will contain results from other softwares such as: **protein_match**, **signalpeptides**, **transmembrane**, **Virulence Factors**, **Operon**, **ncRNA**
- Apply bedtools sort with start sites and bedtools merge for overlaps.

###### For final results the following files are present

- Clustered files (fasta/uc)
- Trimmed(if needed) interproscan/eggnog/ard/operon results with cluster
- Original signalP/tmhmm results
- Original ncRNA gffs and CRISPR results
- Merged gff files (for each sample) with two types
- 50 one-line annotation gff files with gene names (for each sample)
